src/attribute.rs

```diff
@@ -520,37 +521,191 @@ impl CK_ATTRIBUTE {
             AttrType::IgnoreType => Ok(from_ignore(self.type_, None)),
         }
     }
+}
 
-    pub fn from_slice<'a>(
-        type_: CK_ATTRIBUTE_TYPE,
-        value: &'a [u8],
-    ) -> CK_ATTRIBUTE {
-        CK_ATTRIBUTE {
-            type_: type_,
-            pValue: value.as_ptr() as *mut std::ffi::c_void,
-            ulValueLen: value.len() as CK_ULONG,
+#[derive(Debug)]
+pub struct CkAttrs<'a> {
+    v: Vec<Cow<'a, [u8]>>,
+    p: Cow<'a, [CK_ATTRIBUTE]>,
+    pub zeroize: bool,
+}
+
+impl Drop for CkAttrs<'_> {
+    fn drop(&mut self) {
+        if self.zeroize {
+            while let Some(mut elem) = self.v.pop() {
+                elem.to_mut().zeroize();
+            }
         }
     }
+}
 
-    pub fn from_ulong<'a>(
-        type_: CK_ATTRIBUTE_TYPE,
-        value: &'a CK_ULONG,
-    ) -> CK_ATTRIBUTE {
-        CK_ATTRIBUTE {
-            type_: type_,
-            pValue: value as *const CK_ULONG as *mut std::ffi::c_void,
-            ulValueLen: sizeof!(CK_ULONG),
+impl<'a> CkAttrs<'a> {
+    pub fn new() -> CkAttrs<'static> {
+        Self::with_capacity(0)
+    }
+
+    pub fn with_capacity(capacity: usize) -> CkAttrs<'static> {
+        CkAttrs {
+            v: Vec::new(),
```

Comment:
should this vector have the default capacity too?

Reply:
Not really, in most cases you should not really add anything in it and instead just depend on lifetimes passing in references that do not need to be copied at all.

Only the add_owned_something() function need that vector, and I tried hard to avoid them most of the time as it is inefficient to copy stuff all the time when a simple pointer is enough.

The problem was making sure Rust catches when we are referencing a pointer so it won't free it before CkAttrs is, which the whole lifetime on the struct and the impl and the functions takes care of now that the raw CK_ATTRIBUTES slice could not.